src/adaXT/random_forest/RandomForest.py

```diff
@@ -0,0 +1,355 @@
+
+        self.predict_values = SharedNumpyArray(X)
+
+        # Predict using all the trees, each column is the predictions from one
+        # tree
+        tree_predictions = self.__predict_trees()
+        self.predict_values.unlink()
+
+        if self.forrest_type == "Regression":
+            # Return the mean answer from all trees for each row
+            return np.mean(tree_predictions, axis=1)
+
+        elif self.forrest_type == "Classification":
+            # Return the element most voted for by trees for each row
+            # QUESTION: how to handle equal majority vote?
+            return np.apply_along_axis(
+                self.most_frequent_element, 1, tree_predictions)
+
+    # Function used to find the most frequent element of an array
+    def most_frequent_element(self, arr):
+        values, counts = np.unique(arr, return_counts=True)
+        return values[np.argmax(counts)]
+
+    def predict_proba(self, X: np.ndarray):
+        """
+        Predicts a probability for each response for given X values using the trees of the forrest
+
+        Parameters
+        ----------
+        X : np.ndarray
+            (N, M) numpy array with features to predict
+
+        Returns
+        -------
+        np.ndarray
+            Returns an ndarray with the probabilities for each class per observation in X. The order of the classes corresponds to that in the attribute classes
+        """
+        if not self.forest_fitted:
+            raise AttributeError(
+                "The forrest has not been fitted before trying to call predict_proba")
+
+        # Make sure that predict_proba is only called on Classification
+        # forrests
+        if self.forrest_type != "Classification":
+            raise ValueError(
+                "predict_proba can only be called on a Classification tree")
+
+        # Check dimensions
+        self.check_dimensions(X)
+
+        self.predict_values = SharedNumpyArray(X)
+        # Predict_proba using all the trees, each element of list is the
+        # predict_proba from one tree
+        tree_predictions = self.__predict_proba_trees()
+        self.predict_values.unlink()
+
+        # Stack the predict_probas
+        stacked_tree_predictions = np.stack(tree_predictions, axis=0)
+
+        # Return the mean along the newly created axis
+        return np.mean(stacked_tree_predictions, axis=0)
+
+    def check_dimensions(self, X: np.ndarray):
+        # If there is only a single point
+        if X.ndim == 1:
+            if (X.shape[0] != self.n_features):
+                raise ValueError(
+                    f"Number of features should be {self.n_features}, got {X.shape[0]}")
+        else:
+            if X.shape[1] != self.n_features:
+                raise ValueError(
+                    f"Dimension should be {self.n_features}, got {X.shape[1]}")
+
+    def check_input(self, X: np.ndarray, Y: np.ndarray):
```

Comment:
If this function already checks the input of the data, then we
should have a switch in the Decision tree, such that the singular
tree doesn't also check the input. Would be double work, and could
be a small time wasted.

Reply:
Done.